### PR TITLE
Initial commit based off of interceptor

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

File: proxy.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/getlantern/ops"
+)
+
+var (
+	log = golog.LoggerFor("proxy")
+)
+
+type DialFunc func(network, addr string) (conn net.Conn, err error)
+
+type Interceptor func(op ops.Op, w http.ResponseWriter, req *http.Request)
```
Comment (joesis) [inline proxy.go]: It feels to me it's unnecessary to introduce dependency to `ops` package which is more suitable at the application level. How about `func(w http.ResponseWriter, req *http.Request) error`, and let the caller decide what to do with the error?

File: proxy_connect.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+// request
+//
+// idleTimeout: if specified, lets us know to include an appropriate
+// KeepAlive: timeout header in the CONNECT response
+//
+// bufferSource: specifies a BufferSource, leave nil to use default
+//
+// dial: the function that's used to dial upstream
+func CONNECT(
+	defaultPort int,
```
Comment (joesis) [inline proxy_connect.go]: It's also what the standard library do https://golang.org/src/net/http/transport.go#L1933

File: proxy.go (hunk 1 above)
Comment (oxtoacart) [inline proxy.go]: Yes, now that we're not pipelining anymore that makes sense, as there'll always be just one error.

File: proxy_connect.go (hunk 2 above)
Comment (oxtoacart) [inline proxy_connect.go]: Oh, good point. I shouldn't even be using the HOST header. See discussion here:

https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number

Comment (oxtoacart) [inline proxy_connect.go]: Fixed

File: proxy_connect.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+// BufferSource is a source for buffers used in reading/writing.
+type BufferSource interface {
+	Get() []byte
+	Put(buf []byte)
+}
+
+// CONNECT returns a Handler that processes CONNECT requests.
+//
+// defaultPort: specifies which port to use if none was found in the initial
```
Comment (joesis) [inline proxy_connect.go]: comment can be removed.

Comment (oxtoacart) [inline proxy_connect.go]: Fixed.

File: proxy_connect.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	}
+	closeDownstream = true
+
+	// Note - for CONNECT requests, we use the Host from the request URL, not the
+	// Host header. See discussion here:
+	// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
+	upstream, err = ic.dial("tcp", req.URL.Host)
+	if err != nil {
+		fullErr := errors.New("Unable to dial upstream: %s", err)
+		ic.respondBadGatewayHijacked(downstream, req, fullErr)
```
Comment (joesis) [inline proxy_connect.go]: Maybe not worth doing it, but if we swap the order of hijacking client connection and dialing origin, we can use `respondBadGateway` here and remove `respondBadGatewayHijacked` altogether. `Hijack` operation rarely fail, so it won't waste time to dial origin in most cases.

Comment (oxtoacart) [inline proxy_connect.go]: Great idea, done!

File: proxy_http.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+			return nil
+		}
+
+		// Preserve remote address from original request
+		req.RemoteAddr = remoteAddr
+		first = false
+	}
+}
+
+func (ic *httpInterceptor) writeResponse(downstream net.Conn, resp *http.Response) error {
```
Comment (oxtoacart) [inline proxy_http.go]: Changed type to `io.Writer`, kept name of `downstream` because `w` is used elsewhere to refer to the `ResponseWriter`.

**Comment by joesis** [comment]
Ok, that's all I could find @oxtoacart . Cool stuff, as always. We don't need to actually merge right?

**Comment by oxtoacart** [comment]
@joesis Right, it's "pre-merged". Thanks for the review, great improvements as always!

**Comment by myleshorton** [comment]
Should this be merged @oxtoacart and @joesis? Or closed?

**Comment by oxtoacart** [comment]
Closed, this was just for illustration.

**Pull request closed without merging**